File: app/api/v1/routes/budget.py
from typing import Optional

from fastapi import APIRouter, HTTPException, Response
from pydantic import BaseModel

from app.api.deps import CurrentUser, SessionDep
from app.models.budget_expense import BudgetExpense
from app.models.trip import Trip
from app.schemas.budget import BudgetExpenseCreate, BudgetExpenseUpdate, BudgetExpenseResponse, BudgetResponse

router = APIRouter()


class BudgetLimitUpdate(BaseModel):
    limit: Optional[float] = None


def _get_trip(trip_id: int, user_id: int, db: SessionDep) -> Trip:
    trip = db.get(Trip, trip_id)
    if not trip or trip.user_id != user_id:
        raise HTTPException(status_code=404, detail="Trip not found")
    return trip


@router.get("/", response_model=BudgetResponse)
def get_budget(trip_id: int, db: SessionDep, current_user: CurrentUser):
    trip = _get_trip(trip_id, current_user.id, db)
    expenses = db.query(BudgetExpense).filter(BudgetExpense.trip_id == trip_id).all()
    return BudgetResponse(limit=trip.budget_limit, expenses=expenses)


@router.patch("/limit", response_model=BudgetResponse)
def update_budget_limit(
    trip_id: int, body: BudgetLimitUpdate, db: SessionDep, current_user: CurrentUser
):
    trip = _get_trip(trip_id, current_user.id, db)
    trip.budget_limit = body.limit
    db.commit()
    db.refresh(trip)
    expenses = db.query(BudgetExpense).filter(BudgetExpense.trip_id == trip_id).all()
    return BudgetResponse(limit=trip.budget_limit, expenses=expenses)


@router.post("/expenses", response_model=BudgetExpenseResponse, status_code=201)
def create_expense(
    trip_id: int, expense_in: BudgetExpenseCreate, db: SessionDep, current_user: CurrentUser
):
    _get_trip(trip_id, current_user.id, db)
    expense = BudgetExpense(
        trip_id=trip_id,
        label=expense_in.label,
        amount=expense_in.amount,
        category=expense_in.category,
    )
    db.add(expense)
    db.commit()
    db.refresh(expense)
    return expense


@router.patch("/expenses/{expense_id}", response_model=BudgetExpenseResponse)
def update_expense(
    trip_id: int,
    expense_id: int,
    expense_in: BudgetExpenseUpdate,
    db: SessionDep,
    current_user: CurrentUser,
):
    _get_trip(trip_id, current_user.id, db)
    expense = db.get(BudgetExpense, expense_id)
    if not expense or expense.trip_id != trip_id:
        raise HTTPException(status_code=404, detail="Expense not found")
    if expense_in.label is not None:
        expense.label = expense_in.label
    if expense_in.amount is not None:
        expense.amount = expense_in.amount
    if expense_in.category is not None:
        expense.category = expense_in.category
    db.commit()
    db.refresh(expense)
    return expense


@router.delete("/expenses/{expense_id}", status_code=204)
def delete_expense(
    trip_id: int, expense_id: int, db: SessionDep, current_user: CurrentUser
):
    _get_trip(trip_id, current_user.id, db)
    expense = db.get(BudgetExpense, expense_id)
    if not expense or expense.trip_id != trip_id:
        raise HTTPException(status_code=404, detail="Expense not found")
    db.delete(expense)
    db.commit()
    return Response(status_code=204)

File: ui/src/shared/api/budget.ts
import { API_URL } from '../../app/config';

export interface BudgetExpense {
  id: number;
  trip_id: number;
  label: string;
  amount: number;
  category: string;
  created_at: string;
}

export interface BudgetData {
  limit: number | null;
  expenses: BudgetExpense[];
}

const base = (tripId: number) => `${API_URL}/v1/trips/${tripId}/budget`;

const authHeaders = (token: string) => ({
  Authorization: `Bearer ${token}`,
  'Content-Type': 'application/json',
});

export async function getBudget(token: string, tripId: number): Promise<BudgetData> {
  const res = await fetch(`${base(tripId)}/`, { headers: authHeaders(token) });
  if (!res.ok) throw new Error(`Failed to load budget (${res.status})`);
  return res.json();
}

export async function updateBudgetLimit(token: string, tripId: number, limit: number | null): Promise<BudgetData> {
  const res = await fetch(`${base(tripId)}/limit`, {
    method: 'PATCH',
    headers: authHeaders(token),
    body: JSON.stringify({ limit }),
  });
  if (!res.ok) throw new Error(`Failed to update budget limit (${res.status})`);
  return res.json();
}

export async function createExpense(
  token: string,
  tripId: number,
  expense: { label: string; amount: number; category: string },
): Promise<BudgetExpense> {
  const res = await fetch(`${base(tripId)}/expenses`, {
    method: 'POST',
    headers: authHeaders(token),
    body: JSON.stringify(expense),
  });
  if (!res.ok) throw new Error(`Failed to add expense (${res.status})`);
  return res.json();
}

export async function updateExpense(
  token: string,
  tripId: number,
  expenseId: number,
  patch: { label?: string; amount?: number; category?: string },
): Promise<BudgetExpense> {
  const res = await fetch(`${base(tripId)}/expenses/${expenseId}`, {
    method: 'PATCH',
    headers: authHeaders(token),
    body: JSON.stringify(patch),
  });
  if (!res.ok) throw new Error(`Failed to update expense (${res.status})`);
  return res.json();
}

export async function deleteExpense(token: string, tripId: number, expenseId: number): Promise<void> {
  const res = await fetch(`${base(tripId)}/expenses/${expenseId}`, {
    method: 'DELETE',
    headers: authHeaders(token),
  });
  if (!res.ok) throw new Error(`Failed to delete expense (${res.status})`);
}
